File: src/my_py_pkg/my_py_pkg/minimal_publisher.py
import rclpy
from rclpy.executors import ExternalShutdownException
from rclpy.node import Node

from example_interfaces.msg import String


class MinimalPublisher(Node):
    """Publisher node that publishes a string message."""

    def __init__(self):
        super().__init__("minimal_publisher")
        # Create publisher object
        self._publisher = self.create_publisher(String, "topic", 10)

        # Create timer object
        self._timer = self.create_timer(0.5, self._timer_callback)

        # Counter variable to keep track of the number of messages published
        self._counter = 0

    def _timer_callback(self):
        msg = String()
        msg.data = "Hello World: %d" % self._counter
        self._publisher.publish(msg)

        self.get_logger().info('Publishing: "%s"' % msg.data)
        self._counter += 1


def main(args=None):
    try:
        rclpy.init(args=args)
        node = MinimalPublisher()
        rclpy.spin(node)
    except (KeyboardInterrupt, ExternalShutdownException):
        pass
    finally:
        if node is not None:
            node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/my_cpp_pkg/src/minimal_publisher.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/string.hpp"

/**
 * Publisher example that periodically publishes a message.
 */
class MinimalPublisher : public rclcpp::Node
{
    public:

    /**
     * Constructor (call Node class constructor with node name)
     */
     MinimalPublisher() : Node("minimal_publisher")
     {
         // Initialize publisher
         publisher_ = this->create_publisher<example_interfaces::msg::String>("my_topic", 10);

         // Initialize timer
         timer_ = this->create_wall_timer(
             std::chrono::milliseconds(500),
             std::bind(&MinimalPublisher::timer_callback, this));

         // Counter for messages sent
         counter_ = 0;
     }

     private:
        /**
         * Timer callback function
         */
        void timer_callback()
        {
            auto msg = example_interfaces::msg::String();
            msg.data = "Hello, world!: " + std::to_string(counter_);
            // Publish message
            publisher_->publish(msg);

            // Log message
            RCLCPP_INFO(this->get_logger(), "Published message %s", msg.data.c_str());

            // Increment counter
            counter_++;
        }

        rclcpp::Publisher<example_interfaces::msg::String>::SharedPtr publisher_;
        rclcpp::TimerBase::SharedPtr timer_;
        size_t counter_;
};


/**
 * Main entrypoint
 */
int main(int argc, char * argv[])
{
    // Initialize ROS2
    rclcpp::init(argc, argv);

    // Initialize and run node
    auto node = std::make_shared<MinimalPublisher>();
    rclcpp::spin(node);

    // Cleanup
    rclcpp::shutdown();

    return 0;
}
